src/main.rs

```diff
@@ -11,34 +11,58 @@ const NO_CACHE_TTL: i32 = -1;
 /// be used to route based on the request properties (such as method or path),
 /// send the request to a backend, make completely new requests and/or generate
 /// synthetic responses.
-fn handle_request(req: Request<Body>) -> Result<Response<Body>, Error> {
+fn handle_request(mut req: Request<Body>) -> Result<Response<Body>, Error> {
+
+    // Make any desired changes to the client request
+    req.headers_mut().insert("Host", HeaderValue::from_static("example.com"));
+
     // Pattern match on the request method and path.
     match (req.method(), req.uri().path()) {
-        // If request is `GET /`.
+
+        // If request method is not GET or HEAD, error
+        (method, _) if !(vec![Method::HEAD, Method::GET].contains(method)) => Ok(
```

Comment:
Whilst I like your solution, you can avoid creating the anonymous vector with an OR operator in your guard conditional. Such as:

```
(method, _) if method != Method::GET || method != Method::HEAD => Ok(...)
```
It's not as clean or terse as yours, but avoids the memory assignment. 

Another way would be to use "Or patterns", which are still experimental, such as:
```
(Method::POST | Method::PUT, _) => Ok(...)
```

/ cc @acfoltzer @katie-martin-fastly for suggestions / thoughts

Reply:
Good catch on the point that arbitrarily nested or patterns are [still experimental](https://github.com/rust-lang/rfcs/blob/master/text/2535-or-patterns.md).

I think I would lean towards the first suggestion in your comment above, @phamann, for two reasons:

* Routing _is_ listed as a use case in the or-pattern RFC, but I think introducing the cognitive load of [enabling unstable features](https://doc.rust-lang.org/unstable-book/language-features/or-patterns.html) on top of the existing "_how does Rust work?_" context is asking a lot from users.
* Judging by the [tracking issue](https://github.com/rust-lang/rust/issues/54883), there is also still some work to be done before this feature stabilizes. If there's another way, i.e. using `if foo || bar =>`, then that's probably best for now.

Reply:
You could also apply the distributive property to get a top-level or-pattern:
```suggestion
        (Method::HEAD, _) | (Method::GET, _) => Ok(
```
I think I'd favor this because pattern matching is preferred in general to `==`.

Reply:
... I just realized I was missing a negation. I would actually structure this differently. Above the current fallthrough case, I would add a `(Method::GET, path) =>` case that gives the 404 response. Then, the fallthrough case must be an unsupported method, so you could return the appropriate error.

Beyond resolving the immediate syntactic questions, this also is more defensive against future changes to the `Method` type.

Reply:
So the reason I'd prefer to keep my proposed code here is that it is more extensible. The inline vector scales better if I were to add more methods, or if I wanted to pull out the list of methods into a constant or a lookup.  Is there a better way to suggest that potential?

Reply:
So, my main objection here is not the use of a list of methods (though I would recommend using a slice `&[Method::HEAD, Method::GET]` rather than a vec to avoid allocation), but that logically this is expressing a fallthrough case, and so in a Rust program I would expect to see it as a fallthrough pattern.

Perhaps this could be transformed to a clause that matches if the method is a member of a list, as opposed to ensuring that the list doesn't contain the method?

Reply:
Abstracted this out.  What do you think?

Reply:
Whilst I like your new abstraction, I'm confused as it isn't the same logic as before. It seems to be allowing all methods not just a subset i.e. before you were only allowing GET and HEAD.

Reply:
Customer will customise so it doesn't matter that much, but the new impl was inconsistent with the error msg so I updated the message.

Reply:
Ok. My only concern with your new list is it allows ALL methods and therefore a strange example, as if a client sends a request with an unknown or malformed method it won't even be forwarded by h2o to xqd and therefore makes your logic redundant. Could we at least reduce the allow list to make a better/more realistic example?